main/parsers/zettle_parser.py

```diff
@@ -1,71 +1,89 @@
 al = AssetLoader()
 
 
 class ZettleParser:  # Parser
     def __init__(self):
-        self.zettle = self.set_cred()
+        self.auth_token = None
 
     def intakt_type(self):
         return "Zettle"
 
-    def set_cred(self):
-        access_file = (
+    def get_new_auth_token(self):
+        access_file_path = (
             os.path.dirname(os.path.realpath(__file__)) + "/../credentials/access.json"
         )
+        with open(access_file_path) as access_file:
+            access_json = json.load(access_file)
 
-        with open(access_file) as f:
-            access_cred = json.load(f)
-            f.close()
+        token_response = requests.post(
+            "https://oauth.zettle.com/token",
+            headers={"Content-Type": "application/x-www-form-urlencoded"},
+            data=f"grant_type={access_json['grant_type']}&client_id={access_json['client_id']}&assertion={access_json['assertion']}",
+        )
+
+        if token_response.ok:
+            response_json = token_response.json()
+            self.auth_token = response_json["access_token"]
+            response_json["expires_at"] = time.time() + response_json["expires_in"]
 
-        client_id = access_cred["client_id"]
-        redirect_uri = "https://httpbin.org/get"
+            saved_token_file_path = (
+                os.path.dirname(os.path.realpath(__file__))
+                + "/../credentials/last_used_zettle_token.json"
+            )
+            with open(saved_token_file_path, "w") as saved_token_file:
+                json.dump(response_json, saved_token_file)
+        else:
+            print(
+                f"error, could not get auth token. http response: {token_response.status_code} {token_response.reason}"
+            )
 
-        authorization_base_url = "https://oauth.zettle.com/authorize"
-        scope = ["READ:PURCHASE"]
+    def set_auth_token(self):
+        saved_token_file_path = (
+            os.path.dirname(os.path.realpath(__file__))
+            + "/../credentials/last_used_zettle_token.json"
+        )
 
-        zettle = OAuth2Session(client_id, scope=scope, redirect_uri=redirect_uri)
-        authorization_url, _ = zettle.authorization_url(authorization_base_url)
+        try:
+            with open(saved_token_file_path) as saved_token_file:
+                saved_token_obj = json.load(saved_token_file)
 
-        # TODO: can this be done without human interaction?
-        redirect_response = input(authorization_url + "\n")
+            if time.time() < saved_token_obj["expires_at"]:
+                self.auth_token = saved_token_obj["access_token"]
+                return
 
-        token_url = "https://oauth.zettle.com/token"
-        client_secret = access_cred["client_secret"]
-        zettle.fetch_token(
-            token_url,
-            include_client_id=True,
-            client_secret=client_secret,
-            authorization_response=redirect_response,
-        )
+        except FileNotFoundError:
+            pass
```

Comment:
Same as above, add `exit(0)`

Reply:
We don't want to exit here. The `FileNotFoundError` just means there is no saved token, so we ignore the error and get a new token on the next line.

Reply:
Yes, but `self.auth_token` would not be set right?

Reply:
If `last_used_zettle_token.json` can't be opened, or the token has expired, `self.get_new_auth_token()` gets called and there `self.auth_token` will be set (or it fails and the program exits).

Reply:
Alright, then everything is fine!